internals/cryptutil/crypt_cgo.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+//go:build cgo && linux
+
+package cryptutil
+
+//#cgo pkg-config: libcrypt
+//#include <crypt.h>
+//#include <string.h>
+//#include <malloc.h>
+//struct crypt_data *alloc_cryptdata() {
+//  struct crypt_data *d = malloc(sizeof(struct crypt_data));
+//  if(d == 0) {
+//    return 0;
+//  }
+//  memset(d, 0, sizeof(struct crypt_data));
+//  return d;
+//}
```

Comment:
There's `crypt_ra()` to do this, specifically so that applications don't `malloc()` for sensitive data:

https://manpages.debian.org/experimental/libcrypt1-dev/crypt_r.3.en.html